File: python/gherkin/dialect.py
from __future__ import annotations

import os
import json
from typing import TypedDict

from typing_extensions import Self

DIALECT_FILE_PATH = os.path.join(
    os.path.dirname(__file__),
    "gherkin-languages.json",
)

DialectSpec = TypedDict(
    "DialectSpec",
    {
        "and": list[str],
        "background": list[str],
        "but": list[str],
        "examples": list[str],
        "feature": list[str],
        "given": list[str],
        "rule": list[str],
        "scenario": list[str],
        "scenarioOutline": list[str],
        "then": list[str],
        "when": list[str],
    },
)

with open(DIALECT_FILE_PATH, encoding="utf-8") as file:
    DIALECTS: dict[str, DialectSpec] = json.load(file)


class Dialect:

    @classmethod
    def for_name(cls, name: str) -> Self | None:
        return cls(DIALECTS[name]) if name in DIALECTS else None

    def __init__(self, spec: DialectSpec) -> None:
        self.spec = spec

    @property
    def feature_keywords(self) -> list[str]:
        return self.spec["feature"]

    @property
    def rule_keywords(self) -> list[str]:
        return self.spec["rule"]

    @property
    def scenario_keywords(self) -> list[str]:
        return self.spec["scenario"]

    @property
    def scenario_outline_keywords(self) -> list[str]:
        return self.spec["scenarioOutline"]

    @property
    def background_keywords(self) -> list[str]:
        return self.spec["background"]

    @property
    def examples_keywords(self) -> list[str]:
        return self.spec["examples"]

    @property
    def given_keywords(self) -> list[str]:
        return self.spec["given"]

    @property
    def when_keywords(self) -> list[str]:
        return self.spec["when"]

    @property
    def then_keywords(self) -> list[str]:
        return self.spec["then"]

    @property
    def and_keywords(self) -> list[str]:
        return self.spec["and"]

    @property
    def but_keywords(self) -> list[str]:
        return self.spec["but"]

File: go/dialect.go
package gherkin

import (
	"sort"
	"sync"

	messages "github.com/cucumber/messages/go/v24"
)

type Dialect struct {
	Language     string
	Name         string
	Native       string
	Keywords     map[string][]string
	KeywordTypes map[string]messages.StepKeywordType

	stepKeywordBuilder sync.Once
	stepKeywordsCache  []string
}

func (g *Dialect) FeatureKeywords() []string {
	return g.Keywords["feature"]
}

func (g *Dialect) RuleKeywords() []string {
	return g.Keywords["rule"]
}

func (g *Dialect) ScenarioKeywords() []string {
	return g.Keywords["scenario"]
}

func (g *Dialect) StepKeywords() []string {
	// This only needs to be calculated the once, not every time
	g.stepKeywordBuilder.Do(func() {
		result := g.Keywords["given"]
		result = append(result, g.Keywords["when"]...)
		result = append(result, g.Keywords["then"]...)
		result = append(result, g.Keywords["and"]...)
		result = append(result, g.Keywords["but"]...)

		sort.Slice(result, func(i, j int) bool {
			return result[i] > result[j]
		})
		g.stepKeywordsCache = result
	})

	return g.stepKeywordsCache
}

func (g *Dialect) BackgroundKeywords() []string {
	return g.Keywords["background"]
}

func (g *Dialect) ScenarioOutlineKeywords() []string {
	return g.Keywords["scenarioOutline"]
}

func (g *Dialect) ExamplesKeywords() []string {
	return g.Keywords["examples"]
}

func (g *Dialect) StepKeywordType(keyword string) messages.StepKeywordType {
	return g.KeywordTypes[keyword]
}

type DialectProvider interface {
	GetDialect(language string) *Dialect
}

type gherkinDialectMap map[string]*Dialect

func (g gherkinDialectMap) GetDialect(language string) *Dialect {
	return g[language]
}
